lib/types.nix

```diff
@@ -367,22 +367,47 @@ rec {
         name = "coercedTo";
         description = "${finalType.description} or ${coercedType.description}";
         check = x: finalType.check x || coercedType.check x;
-        merge = loc: defs:
+        merge =
           let
             coerceVal = val:
               if finalType.check val then val
               else let
                 coerced = coerceFunc val;
               in assert finalType.check coerced; coerced;
 
-          in finalType.merge loc (map (def: def // { value = coerceVal def.value; }) defs);
+          in loc: defs: finalType.merge loc (map (def: def // { value = coerceVal def.value; }) defs);
         getSubOptions = finalType.getSubOptions;
         getSubModules = finalType.getSubModules;
         substSubModules = m: coercedTo coercedType coerceFunc (finalType.substSubModules m);
         typeMerge = t1: t2: null;
         functor = (defaultFunctor name) // { wrapped = finalType; };
       };
 
+    # Function from `from` to `to`.
+    function = from: to:
+      assert from.getSubModules == null;
+      mkOptionType rec {
+        name = "function";
```

Comment:
@nbp Won't this be a heavy computation (zipping two package sets)? Still yes, sans that it sounds like a good idea. I'll try to implement it and see if it would indeed be slow (I think it won't be, after all).